src/main/java/net/snowflake/client/core/QueryContextCache.java

```diff
@@ -137,163 +135,218 @@ public void clearCache() {
   }
 
   /**
-   * Deserialize the query context from the base64 encoded arrow data.
-   *
-   * @param data the base64 encoded query context that was serialized using Arrow.
-   */
-  public void deserializeFromArrowBase64(String data) {
-    synchronized (this) {
-      logger.debug("deserializeFromArrowBase64() called: data from the server = {}", data);
+  * @param data: the QueryContext Object serialized as a JSON format string
+  */
+  public void deserializeQueryContextJson(String data) {
 
+    synchronized (this) {
       // Log existing cache entries
       logCacheEntries();
 
       if (data == null || data.length() == 0) {
         // Clear the cache
         clearCache();
 
-        logger.debug("deserializeFromArrowBase64() returns");
         // Log existing cache entries
         logCacheEntries();
 
         return;
       }
 
-      byte[] decoded = Base64.getDecoder().decode(data);
-      ByteArrayInputStream input = new ByteArrayInputStream(decoded);
-      try (ArrowStreamReader reader = new ArrowStreamReader(input, new RootAllocator())) {
-        while (reader.loadNextBatch()) {
-          VectorSchemaRoot root = reader.getVectorSchemaRoot();
-          for (int i = 0; i < root.getRowCount(); ++i) {
-            QueryContextElement qce = deserializeEntry(root, i);
-            // Merge the element in the existing cache
-            merge(qce.id, qce.readTimestamp, qce.priority, qce.context);
+    try{
+      JsonNode rootNode = jsonObjectMapper.readTree(data);
+
+      // Deserialize the main entry. An example JSON is:
+      // {
+      //   "main_entry": {
+      //     "id": 1,
+      //     "read_timestamp": 123456789,
+      //     "priority": 0,
+      //     "context": "base64 encoded context"
+      // }
+      JsonNode mainEntryNode = rootNode.path("main_entry");
+      if (!mainEntryNode.isMissingNode()) {
+        QueryContextElement mainEntry = deserializeQueryContextElement(mainEntryNode);
+        merge(mainEntry.id, mainEntry.readTimestamp, mainEntry.priority, mainEntry.context);
+      }
+
+      // Deserialize the entries. An example JSON is:
+      // {
+      //   "entries": [
+      //     {
+      //       "id": 2,
+      //       "read_timestamp": 123456789,
+      //       "priority": 1,
+      //       "context": "base64 encoded context"
+      //     },
+      //     {
+      //       "id": 3,
+      //       "read_timestamp": 123456789,
+      //       "priority": 2,
+      //       "context": "base64 encoded context"
+      //     }
+      //   ]
+
+      JsonNode entriesNode = rootNode.path("entries");
+      if (entriesNode != null && entriesNode.isArray()) {
```

Comment:
why do we need two different kinds of entries? `mainEntry` `entries`? If I understand correctly `mainEntry` is for main metadata store ( Active Invalidation effort) can we not have all of them under a single `querycontext` element and ensure the presence of the main metastore entry by always setting its priority as 0?

Reply:
I change the `QueryContextDTO` class to contain only `entries`. The JSON received from GS still has `main_entry` and `entries` field.